### DISCUSSION TITLE
Unable to create the messaging service without a project ID

### DISCUSSION THREAD

**Discussion by marten66** [kreait/firebase-php#554] [Q&A] [answered]
I got this Error, when i wan't to send a CloudMessage: 
Fatal error: Uncaught Kreait\Firebase\Exception\RuntimeException: Unable to create the messaging service without a project ID in D:\wwwroot\myProject\www\vendor\kreait\firebase-php\src\Firebase\Factory.php:492 

I think my project ID is in the firebase_credentials.json file, witch seems to be accessable. 
What is the problem?

My code: 
require '../vendor/autoload.php';

use Kreait\Firebase\Factory;
use Kreait\Firebase\Messaging\CloudMessage;
use Kreait\Firebase\Messaging\Message;

$factory = (new Factory)->withServiceAccount('../../firebase/firebase_credentials.json');

$factory = (new Factory())
    ->withDatabaseUri('https://myProject-default-rtdb.firebaseio.com');

$messaging = $factory->createMessaging();    

$deviceToken = "12345";    

 $notification = [
        "body" => "test php",
        "title" => "test from php"
    ];

$message = CloudMessage::withTarget('token', $deviceToken)
    ->withNotification($notification) // optional
    //->withData($data) // optional
;

$messaging->send($message);

**Comment by jeromegamez** [comment]
Can you confirm that the JSON file has a `"type": "service_account"` in it? If not, please generate Service Account credentials as described on https://firebase-php.readthedocs.io/en/5.14.1/setup.html

**Reply by marten66** [reply]
Yes there is. This is my json file: 
{
  "type": "service_account",
  "project_id": "myProject",
  "private_key_id": "xxxxxxxxxxxxxxxx",
  "private_key": "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx",
  "client_email": "firebase-adminsdk-ls4xp@myProject.iam.gserviceaccount.com",
  "client_id": "xxxxxxxxxxxxxxxxxxxxx",
  "auth_uri": "https://accounts.google.com/o/oauth2/auth",
  "token_uri": "https://oauth2.googleapis.com/token",
  "auth_provider_x509_cert_url": "https://www.googleapis.com/oauth2/v1/certs",
  "client_x509_cert_url": "https://www.googleapis.com/robot/v1/metadata/x509/firebase-adminsdk-ls4xp%40myProject.iam.gserviceaccount.com"
}

**Reply by marten66** [reply]
If i underline the name service_account in the beginning. i got another error message. that tells me, that the program usualy find service_account. 
"type": "_service_account",

Fatal error: Uncaught Kreait\Firebase\Exception\InvalidArgumentException: ../../firebase/firebase_credentials.json could not be parsed to a Service Account: A Service Account specification must have a field "type"

**Reply by jeromegamez** [reply]
I really don't know what's happing here 😅. Your code looks alright, the JSON looks alright, from the SDK perspective everything should be fine. But it's not, and I don't have a clue 😬

**Comment by jeromegamez** [comment, marked as answer]
Now I see it! You are creating the factory twice! The first time you have a service account but no database url, the second time you have a database url but no service account. Since you overwrite the first one with the second one, that's the one you're using.

```php
$factory = (new Factory())
    ->withServiceAccount('...')
    ->withDatabaseUri('...');
```

will work 😅

**Reply by marten66** [reply]
Thanks, that was exactly the problem.
No error so far. Followup question:

1) Can i printout the response from Google firebase, so that i can be sure, message was successfully transmited? 
2) Should i see the message here afterwards (see screenshot)?
Nothing happens. I see only my test messages generated on firebase platform. 

![grafik](https://user-images.githubusercontent.com/60956844/106262717-b4b3ed00-6223-11eb-8051-14df6fb55272.png)

**Reply by jeromegamez** [reply]
As long as you don't get an error when sending the message, you can be sure that the message was successfully transmitted from your environment to Firebase and that Firebase has accepted it. Checking if the message if and when the message was actually _received_ on the client(s) is unfortunately not possible.

Only messages sent from the FCM Web Console are visible in the console - messages sent via an Admin SDK/the REST APIs don't show up there, unfortunately.

**Reply by jeromegamez** [reply]
There's the "Reports/Berichte" tab of course, but it has no details. I haven't tried the BigQuery export yet, so I can't tell you how detailed that would be 😅 

https://firebase.google.com/docs/cloud-messaging/understand-delivery#message-delivery-reports

**Reply by marten66** [reply]
Thx I see, thats fine.
I see that after a dispatch the count is increased. So it works. But unfortunately the messages do not arrive at my app on the mobile.  When I send a test message via the Firebase platform, the push messages arrive. Do you have any idea what the problem could be? 

My app is programmed with Ionic Angular. 

import { Injectable } from '@angular/core';
import {
  Plugins,
  PushNotification,
  PushNotificationToken,
  PushNotificationActionPerformed,
  Capacitor
} from '@capacitor/core';
import { Router } from '@angular/router';
 
const { PushNotifications } = Plugins;
 
@Injectable({
  providedIn: 'root'
})
export class FcmService {
 
  constructor(private router: Router) { }
 
  public initPush() {
    if (Capacitor.platform !== 'web') {
      this.registerPush();
    }
  }
 
  private registerPush() {
    PushNotifications.requestPermission().then((permission) => {
      if (permission.granted) {
        // Register with Apple / Google to receive push via APNS/FCM
        PushNotifications.register();
        //console.log('registerPush has done!');
      } else {
        // No permission for push granted
      }
    });
 
    PushNotifications.addListener(
      'registration',
      (token: PushNotificationToken) => {
        console.log('My token: ' + JSON.stringify(token));
        //alert('My token: ' + JSON.stringify(token));
      }
    );
 
    PushNotifications.addListener('registrationError', (error: any) => {
      console.log('Error: ' + JSON.stringify(error));
    });
 
    PushNotifications.addListener(
      'pushNotificationReceived',
      async (notification: PushNotification) => {
        console.log('Push received: ' + JSON.stringify(notification));
      }
    );
 
    PushNotifications.addListener(
      'pushNotificationActionPerformed',
      async (notification: PushNotificationActionPerformed) => {
        const data = notification.notification.data;
        console.log('Action performed: ' + JSON.stringify(notification.notification));
        if (data.detailsId) {
          this.router.navigateByUrl('/details/${data.detailsId}');
        }
      }
    );
  }
}

**Reply by jeromegamez** [reply]
There are many factors that play into message delivery - I can't go into detail for all of them, but some are whether you're sending to a device directly or to a topic (in that case the device needs to be subscribed to the topic), if it's a notification message or a data message (https://firebase.google.com/docs/cloud-messaging/concept-options#notifications_and_data_messages), how Firebase delivers the messages, how the clients process them… that's all out of scope of what I can provide here unfortunately.

You can [become a sponsor](https://github.com/sponsors/jeromegamez/) and gain access to a Slack Workspace with me and other Firebase aficionados, where we can go into more detail if you wish so ^^

**Reply by marten66** [reply]
I can understand that. Thank you.
